### ISSUE TITLE
Flux finetuning consumes all VRAM(24GB) and shared gpu memeory(32GB) compared to adafactor which uses 20+10GB

### ISSUE THREAD

**Issue by mcmanus82** [LoganBooker/prodigy-plus-schedule-free#22]
Hi

Thanks again for you help in the previous post :) PR for OT implementation is waiting to be merged. I the meanwhile i was using Kohya and Prodigy+ works great for lora's there, but when i try to use it for fine tuning it pretty much eats all the RAM(24GB VRAM + 32 GB shared) while the same settings with Adafactor result in 20GB + 10 GB usage. I tried turning off Factored32 and experimented with multiple settings combinations, none of them seem to have any effect on the memory usage. The fused back hook is found and it says it was successful.

Side issue, i started using the "new" validation loss feature (only for Lora training ATM) added to Kohya and did some testing with Prodigy+. It turns out with the default d0 it cannot converge even after 2000+ steps. When i set the d0 to 1e-5 it converges at around 1700 steps(epoch 36 in the screenshot)

In the attached image the grey run is the default, the blue is the modified d0.
Grey run: 24 image dataset, batch size 4, 1 repeats, use_grams=true, final lr was 5.9e-5
Blue run: 24 image dataset, batch size 4, 2 repeats, use_grams=true, final lr was 1.1e-4

![Image](https://github.com/user-attachments/assets/e3d947f5-edb3-4bd9-8598-68fea0231e73)

**Comment by LoganBooker** [comment]
Hey mcmanus82, will do my best to help.

Adafactor with the default settings does not store a first moment, only the factorised second moment, which is relatively small, so it's memory usage is very difficult to beat out of the box. Prodigy+ on the other hand has the weight of Schedule-Free's underlying momentum, so it has about the same memory footprint as AdamW (or Adafactor with the first moment enabled), plus a bit extra from Prodigy's own state variables. I've had a stab at storing `z` (SF's momentum) in a lower rank, but have not been successful.

You can run Prodigy without the Schedule-Free component, and construct the update much like Adafactor does -- in fact, there's a great draft pull on the Prodigy repo by dxqbYD that implements exactly this (https://github.com/konstmish/prodigy/pull/25), but it's fundamentally not compatible with Prodigy + Schedule-Free together (or if it is, it's beyond my humble abilities).

As for the learning rate, I suspect that's due to the LR dampening I implemented to try and control Prodigy's LR. A side effect is that it can sometimes undershoot the optimal LR. Boosting `d0` can remedy this, but I plan to remove the dampening in a future version. I mention how to modify the optimiser to remove it in this comment (https://github.com/LoganBooker/prodigy-plus-schedule-free/issues/14#issuecomment-2632602223).
